## Исследование объявлений о продаже квартир

В нашем распоряжении данные сервиса Яндекс.Недвижимость — архив объявлений о продаже квартир в Санкт-Петербурге и соседних населённых пунктах за несколько лет.

### Цель проекта

Научиться определять рыночную стоимость объектов недвижимости. Наша задача — установить параметры.
Это позволит построить автоматизированную систему: она отследит аномалии и мошенническую деятельность

Провести следующие исследования:

   -  Изучите следующие параметры: площадь, цена, число комнат, высота потолков. Постройте гистограммы для каждого параметра.
   -  Изучите время продажи квартиры. Постройте гистограмму. Посчитайте среднее и медиану. Опишите, сколько обычно занимает продажа. Когда можно считать, что продажи прошли очень быстро, а когда необычно долго?
   -  Уберите редкие и выбивающиеся значения. Опишите, какие особенности обнаружили.
   -  Какие факторы больше всего влияют на стоимость квартиры? Изучите, зависит ли цена от площади, числа комнат, удалённости от центра. Изучите зависимость цены от того, на каком этаже расположена квартира: первом, последнем или другом. Также изучите зависимость от даты размещения: дня недели, месяца и года.
   -  Выберите 10 населённых пунктов с наибольшим числом объявлений. Посчитайте среднюю цену квадратного метра в этих населённых пунктах. Выделите населённые пункты с самой высокой и низкой стоимостью жилья. Эти данные можно найти по имени в столбце locality_name.
   -  Изучите предложения квартир: для каждой квартиры есть информация о расстоянии до центра. Выделите квартиры в Санкт-Петербурге (locality_name). Ваша задача — выяснить, какая область входит в центр. Создайте столбец с расстоянием до центра в километрах: округлите до целых значений. После этого посчитайте среднюю цену для каждого километра. Постройте график: он должен показывать, как цена зависит от удалённости от центра. Определите границу, где график сильно меняется, — это и будет центральная зона.
   -  Выделите сегмент квартир в центре. Проанализируйте эту территорию и изучите следующие параметры: площадь, цена, число комнат, высота потолков. Также выделите факторы, которые влияют на стоимость квартиры (число комнат, этаж, удалённость от центра, дата размещения объявления). Сделайте выводы. Отличаются ли они от общих выводов по всему городу?

### Описание данных

| Название столбца     | Описание                                                         |
|----------------------|------------------------------------------------------------------|
| airports_nearest     | расстояние до ближайшего аэропорта в метрах (м)                  |
| balcony              | число балконов                                                   |
| ceiling_height       | высота потолков (м)                                              |
| cityCenters_nearest  | расстояние до центра города (м)                                  |
| days_exposition      | сколько дней было размещено объявление (от публикации до снятия) |
| first_day_exposition | дата публикации                                                  |
| floor                | этаж                                                             |
| floors_total         | всего этажей в доме                                              |
| is_apartment         | апартаменты (булев тип)                                          |
| kitchen_area         | площадь кухни в квадратных метрах (м²)                           |
| last_price           | цена на момент снятия с публикации                               |
| living_area          | жилая площадь в квадратных метрах (м²)                           |
| locality_name        | название населённого пункта                                      |
| open_plan            | свободная планировка (булев тип)                                 |
| parks_around3000     | число парков в радиусе 3 км                                      |
| parks_nearest        | расстояние до ближайшего парка (м)                               |
| ponds_around3000     | число водоёмов в радиусе 3 км                                    |
| ponds_nearest        | расстояние до ближайшего водоёма (м)                             |
| rooms                | число комнат                                                     |
| studio               | квартира-студия (булев тип)                                      |
| total_area           | площадь квартиры в квадратных метрах (м²)                        |
| total_images         | число фотографий квартиры в объявлении                           |

Пояснение: апартаменты — это нежилые помещения, которые не относятся к жилому фонду, но имеют необходимые условия для проживания.

In [54]:
# импортируем необходимые библиотеки
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from pymystem3 import Mystem
lemm = Mystem().lemmatize # создаем объект лемматизатора

from nltk import SnowballStemmer
stemm = SnowballStemmer('russian').stem # создаем объект стеммера

pd.options.display.max_columns = None

In [55]:
# загрузим наши данные
data = pd.read_csv('../data/real_estate_data.csv', sep='\t')

# посмотрим информацию о колонках
display(data.head(40))
display(data.info())
display(data.describe())

total_images  last_price  total_area first_day_exposition  rooms  \
0             20  13000000.0      108.00  2019-03-07T00:00:00      3   
1              7   3350000.0       40.40  2018-12-04T00:00:00      1   
2             10   5196000.0       56.00  2015-08-20T00:00:00      2   
3              0  64900000.0      159.00  2015-07-24T00:00:00      3   
4              2  10000000.0      100.00  2018-06-19T00:00:00      2   
5             10   2890000.0       30.40  2018-09-10T00:00:00      1   
6              6   3700000.0       37.30  2017-11-02T00:00:00      1   
7              5   7915000.0       71.60  2019-04-18T00:00:00      2   
8             20   2900000.0       33.16  2018-05-23T00:00:00      1   
9             18   5400000.0       61.00  2017-02-26T00:00:00      3   
10             5   5050000.0       39.60  2017-11-16T00:00:00      1   
11             9   3300000.0       44.00  2018-08-27T00:00:00      2   
12            10   3890000.0       54.00  2016-06-30T00:00:00      2   
13            20   3550000.0       42.80  2017-07-01T00:00:00      2   
14             1   4400000.0       36.00  2016-06-23T00:00:00      1   
15            16   4650000.0       39.00  2017-11-18T00:00:00      1   
16            11   6700000.0       82.00  2017-11-23T00:00:00      3   
17             6   4180000.0       36.00  2016-09-09T00:00:00      1   
18             8   3250000.0       31.00  2017-01-27T00:00:00      1   
19            16  14200000.0      121.00  2019-01-09T00:00:00      3   
20            12   6120000.0       80.00  2017-09-28T00:00:00      3   
21            13   3200000.0       31.60  2018-03-14T00:00:00      1   
22            20   5000000.0       58.00  2017-04-24T00:00:00      2   
23            11   2950000.0       32.00  2016-10-29T00:00:00      1   
24             8   6500000.0       97.20  2015-10-31T00:00:00      2   
25             3   6800000.0       76.00  2015-10-01T00:00:00      2   
26             6   4050000.0       60.00  2017-04-28T00:00:00      4   
27            20   7100000.0       70.00  2017-05-12T00:00:00      3   
28             8   4170000.0       44.00  2017-12-13T00:00:00      1   
29             9   8600000.0      100.00  2016-04-09T00:00:00      3   
30            12   2200000.0       32.80  2018-02-19T00:00:00      1   
31             8   7200000.0       67.90  2017-10-26T00:00:00      2   
32             7   4990000.0       60.00  2016-05-22T00:00:00      3   
33             8   4800000.0       73.00  2018-10-15T00:00:00      4   
34             3   3290000.0       33.00  2018-02-04T00:00:00      1   
35             6  15500000.0      149.00  2017-06-26T00:00:00      5   
36            13   3790000.0       45.00  2017-01-25T00:00:00      2   
37            10   1990000.0       45.80  2017-10-28T00:00:00      2   
38            10   3150000.0       40.00  2018-03-29T00:00:00      1   
39            15   5200000.0       54.40  2018-11-29T00:00:00      2   

    ceiling_height  floors_total  living_area  floor is_apartment  studio  \
0             2.70          16.0        51.00      8          NaN   False   
1              NaN          11.0        18.60      1          NaN   False   
2              NaN           5.0        34.30      4          NaN   False   
3              NaN          14.0          NaN      9          NaN   False   
4             3.03          14.0        32.00     13          NaN   False   
5              NaN          12.0        14.40      5          NaN   False   
6              NaN          26.0        10.60      6          NaN   False   
7              NaN          24.0          NaN     22          NaN   False   
8              NaN          27.0        15.43     26          NaN   False   
9             2.50           9.0        43.60      7          NaN   False   
10            2.67          12.0        20.30      3          NaN   False   
11             NaN           5.0        31.00      4        False   False   
12             NaN           5.0        30.00     

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23699 entries, 0 to 23698
Data columns (total 22 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   total_images          23699 non-null  int64  
 1   last_price            23699 non-null  float64
 2   total_area            23699 non-null  float64
 3   first_day_exposition  23699 non-null  object 
 4   rooms                 23699 non-null  int64  
 5   ceiling_height        14504 non-null  float64
 6   floors_total          23613 non-null  float64
 7   living_area           21796 non-null  float64
 8   floor                 23699 non-null  int64  
 9   is_apartment          2775 non-null   object 
 10  studio                23699 non-null  bool   
 11  open_plan             23699 non-null  bool   
 12  kitchen_area          21421 non-null  float64
 13  balcony               12180 non-null  float64
 14  locality_name         23650 non-null  object 
 15  airports_nearest   

None

total_images    last_price    total_area         rooms  ceiling_height  \
count  23699.000000  2.369900e+04  23699.000000  23699.000000    14504.000000   
mean       9.858475  6.541549e+06     60.348651      2.070636        2.771499   
std        5.682529  1.088701e+07     35.654083      1.078405        1.261056   
min        0.000000  1.219000e+04     12.000000      0.000000        1.000000   
25%        6.000000  3.400000e+06     40.000000      1.000000        2.520000   
50%        9.000000  4.650000e+06     52.000000      2.000000        2.650000   
75%       14.000000  6.800000e+06     69.900000      3.000000        2.800000   
max       50.000000  7.630000e+08    900.000000     19.000000      100.000000   

       floors_total   living_area         floor  kitchen_area       balcony  \
count  23613.000000  21796.000000  23699.000000  21421.000000  12180.000000   
mean      10.673824     34.457852      5.892358     10.569807      1.150082   
std        6.597173     22.030445      4.885249      5.905438      1.071300   
min        1.000000      2.000000      1.000000      1.300000      0.000000   
25%        5.000000     18.600000      2.000000      7.000000      0.000000   
50%        9.000000     30.000000      4.000000      9.100000      1.000000   
75%       16.000000     42.300000      8.000000     12.000000      2.000000   
max       60.000000    409.700000     33.000000    112.000000      5.000000   

       airports_nearest  cityCenters_nearest  parks_around3000  parks_nearest  \
count      18157.000000         18180.000000      18181.000000    8079.000000   
mean       28793.672193         14191.277833          0.611408     490.804555   
std        12630.880622          8608.386210          0.802074     342.317995   
min            0.000000           181.000000          0.000000       1.000000   
25%        18585.000000          9238.000000          0.000000     288.000000   
50%        26726.000000         13098.500000          0.000000     455.000000   
75%        37273.000000         16293.000000          1.000000     612.000000   
max        84869.000000         65968.000000          3.000000    3190.000000   

       ponds_around3000  ponds_nearest  days_exposition  
count      18181.000000    9110.000000     20518.000000  
mean           0.770255     517.980900       180.888634  
std            0.938346     277.720643       219.727988  
min            0.000000      13.000000         1.000000  
25%            0.000000     294.000000        45.000000  
50%            1.000000     502.000000        95.000000  
75%            1.000000     729.000000       232.000000  
max            3.000000    1344.000000      1580.000000

Мы видим что у нас есть пропуски в следующих колонках:
- ceiling_height
- floors_total
- living_area
- is_apartment
- kitchen_area
- balcony
- locality_name
- airports_nearest

В пропусках прослеживаются некоторые зависимости, такие как:
1. В столбцах airports_nearest и cityCenters_nearest имеется прямая зависимость между пропусками данных
2. В столбцах parks_around3000 и parks_nearest также имеется прямая зависимость между пропусками данных, также если в столбце parks_around3000 значение равное 0, то в parks_nearest присутствует пропуск
3. Аналогичная зависимость как в п.2 между столбцами ponds_around3000 и ponds_nearest
4. Пропуски в столбце balcony скорее всего указывают на его отсутствие, их можно заполнить нулями

Займемся заполнением пропусков


In [59]:
# проверим, что если в airports_nearest пропуск, то и в cityCenters_nearest тоже
display(f"Пропусков в колонке airports_nearest: {len(data.query('airports_nearest.isna()'))}")
display(f"Пропусков в колонке cityCenters_nearest: {len(data.query('airports_nearest.isna() and cityCenters_nearest.isna()'))}")

'Пропусков в колонке airports_nearest: 5542'

'Пропусков в колонке cityCenters_nearest: 5518'

Пропуски почти совпадают. Только в 24 записях с пропуском в колонке airports_nearest присутствуют данные в колонке cityCenters_nearest. Зависимость определенно есть, в каких-то случаях пропуски все же заполняются. Стоит передать эту информацию коллегам, собравшим датасет.

Пропуски в airports_nearest можно заполнить средним или медианным значением дальности от аэропорта по каждому городу, но в больших городах, таких как СпБ, удаленность квартиры от аэропорта может измеряться десятками километров, а это может влиять на стоимость квартиры, мы пока не знаем. Для маленьких же городов это было бы корректно, т.к. можем считать, что все квартиры в пределах маленького города примерно равноудалены от аэропортов.

Посчитаем среднюю и медианную удаленность квартир в разных городах, исключив СпБ

In [93]:
data.query('locality_name != "Санкт-Петербург"').groupby('locality_name')['airports_nearest'].agg(['mean', 'median'])

mean  median
locality_name                     
Бокситогорск           NaN     NaN
Волосово               NaN     NaN
Волхов                 NaN     NaN
Всеволожск             NaN     NaN
Выборг                 NaN     NaN
...                    ...     ...
село Путилово          NaN     NaN
село Рождествено       NaN     NaN
село Русско-Высоцкое   NaN     NaN
село Старая Ладога     NaN     NaN
село Шум               NaN     NaN

[363 rows x 2 columns]

Для большинства городов не хватает данных, чтобы посчитать удаленность от аэропорта, потому что по ним совсем нет данных.
Медианная и средняя удаленность у остальных не сильно отличаются, предпочтем заполнить пропуски медианной удаленностью квартир от аэропорта.

In [94]:
data['airports_nearest'] = data.query('locality_name != "Санкт-Петербург"').groupby('locality_name')['airports_nearest'].transform('median')

Для удаленности квартир от центра городов, такой способ не подходит. Поэтому оставим пропуски в столбце cityCenters_nearest как есть.

Займемся пропусками в столбцах parks_around3000 и parks_nearest.
Ранее мы выдвинули предположение, что в столбцах имеется прямая зависимость между пропусками данных, также если в столбце parks_around3000 значение равное 0, то в parks_nearest присутствует пропуск.
Проверим это.

In [96]:
display(f"Пропусков в колонке parks_around3000: {len(data.query('parks_around3000.isna()'))}")
display(f"Пропусков в колонке parks_nearest: {len(data.query('parks_around3000.isna() and parks_nearest.isna()'))}")

'Пропусков в колонке parks_around3000: 5518'

'Пропусков в колонке parks_nearest: 5518'

Действительно наблюдаем прямую зависимость между столбцами. Логично будет предположить, что пропуск в столбце parks_around3000 означает отсутствие парков в радиусе 3км, а значит и расстояние до ближайшего определить не удастся.

Т.о. заполним пропуски в столбце parks_around3000 значением 0, а в колонке parks_nearest оставим пропуски как есть.

In [100]:
data['parks_around3000'].fillna(0, inplace=True)
# проверим отсутствие пропусков в колонке parks_around3000
display(f"Количество пропусков в столбце parks_around3000: {len(data.query('parks_around3000.isna()'))}")

'Количество пропусков в столбце parks_around3000: 0'

Займемся пропусками в колонке is_apartment
Предположим, что пропуск в данной колонке, означает, что квартира не является апартаментами, т.о. заполним пропуски значением False

In [102]:
data['is_apartment'].fillna(False, inplace=True)
# проверим отсутствие пропусков в колонке is_apartment
display(f"Количество пропусков в столбце is_apartment: {len(data.query('is_apartment.isna()'))}")

'Количество пропусков в столбце parks_around3000: 0'

Займемся пропусками в колонке ceiling_height
Предположим, что высота потолков может зависеть от: is_apartment, open_plan или studio
Проверим, есть ли явная зависимость высоты потолков от какого-либо из этих параметров

In [110]:
data.groupby(['is_apartment', 'open_plan', 'studio'])['ceiling_height'].agg('median')

is_apartment  open_plan  studio
False         False      False     2.65
                         True      2.70
              True       False     2.80
True          False      False     3.00
Name: ceiling_height, dtype: float64

Действительно от всех трех параметров есть довольно-таки четкая зависимость высоты потолков
Заполним пропуски полученной медианой

In [112]:
data['ceiling_height'] = data.groupby(['is_apartment', 'open_plan', 'studio'])['ceiling_height'].transform('median')
# проверим отсутствие пропусков в колонке ceiling_height
display(f"Количество пропусков в столбце ceiling_height: {len(data.query('ceiling_height.isna()'))}")

'Количество пропусков в столбце ceiling_height: 0'

Предположим, что пропуск в колонке balcony означает, что в квартире нет балкона, т.о. заполним пропуски значением 0

In [116]:
data['balcony'].fillna(0, inplace=True)
# проверим отсутствие пропусков в колонке ceiling_height
display(f"Количество пропусков в столбце balcony: {len(data.query('balcony.isna()'))}")

'Количество пропусков в столбце balcony: 0'

In [118]:
data.head(50)

total_images  last_price  total_area first_day_exposition  rooms  \
0             20  13000000.0  108.000000  2019-03-07T00:00:00      3   
1              7   3350000.0   40.406250  2018-12-04T00:00:00      1   
2             10   5196000.0   56.000000  2015-08-20T00:00:00      2   
3              0  64900000.0  159.000000  2015-07-24T00:00:00      3   
4              2  10000000.0  100.000000  2018-06-19T00:00:00      2   
5             10   2890000.0   30.406250  2018-09-10T00:00:00      1   
6              6   3700000.0   37.312500  2017-11-02T00:00:00      1   
7              5   7915000.0   71.625000  2019-04-18T00:00:00      2   
8             20   2900000.0   33.156250  2018-05-23T00:00:00      1   
9             18   5400000.0   61.000000  2017-02-26T00:00:00      3   
10             5   5050000.0   39.593750  2017-11-16T00:00:00      1   
11             9   3300000.0   44.000000  2018-08-27T00:00:00      2   
12            10   3890000.0   54.000000  2016-06-30T00:00:00      2   
13            20   3550000.0   42.812500  2017-07-01T00:00:00      2   
14             1   4400000.0   36.000000  2016-06-23T00:00:00      1   
15            16   4650000.0   39.000000  2017-11-18T00:00:00      1   
16            11   6700000.0   82.000000  2017-11-23T00:00:00      3   
17             6   4180000.0   36.000000  2016-09-09T00:00:00      1   
18             8   3250000.0   31.000000  2017-01-27T00:00:00      1   
19            16  14200000.0  121.000000  2019-01-09T00:00:00      3   
20            12   6120000.0   80.000000  2017-09-28T00:00:00      3   
21            13   3200000.0   31.593750  2018-03-14T00:00:00      1   
22            20   5000000.0   58.000000  2017-04-24T00:00:00      2   
23            11   2950000.0   32.000000  2016-10-29T00:00:00      1   
24             8   6500000.0   97.187500  2015-10-31T00:00:00      2   
25             3   6800000.0   76.000000  2015-10-01T00:00:00      2   
26             6   4050000.0   60.000000  2017-04-28T00:00:00      4   
27            20   7100000.0   70.000000  2017-05-12T00:00:00      3   
28             8   4170000.0   44.000000  2017-12-13T00:00:00      1   
29             9   8600000.0  100.000000  2016-04-09T00:00:00      3   
30            12   2200000.0   32.812500  2018-02-19T00:00:00      1   
31             8   7200000.0   67.875000  2017-10-26T00:00:00      2   
32             7   4990000.0   60.000000  2016-05-22T00:00:00      3   
33             8   4800000.0   73.000000  2018-10-15T00:00:00      4   
34             3   3290000.0   33.000000  2018-02-04T00:00:00      1   
35             6  15500000.0  149.000000  2017-06-26T00:00:00      5   
36            13   3790000.0   45.000000  2017-01-25T00:00:00      2   
37            10   1990000.0   45.812500  2017-10-28T00:00:00      2   
38            10   3150000.0   40.000000  2018-03-29T00:00:00      1   
39            15   5200000.0   54.406250  2018-11-29T00:00:00      2   
40             9   3590000.0   36.000000  2017-03-15T00:00:00      1   
41            16   7900000.0   74.000000  2016-05-04T00:00:00      3   
42            13  22000000.0  161.750000  2015-07-08T00:00:00      4   
43            13   9330000.0   48.000000  2017-01-10T00:00:00      2   
44            13   5350000.0   40.000000  2018-11-18T00:00:00      1   
45            17   5200000.0   50.593750  2018-12-02T00:00:00      2   
46            17   6600000.0   52.093750  2019-01-31T00:00:00      2   
47            17   3600000.0   56.093750  2018-10-18T00:00:00      3   
48            10   3600000.0   33.843750  2017-10-03T00:00:00      1   
49             1   3050000.0   30.796875  2018-11-22T00:00:00      1   

    ceiling_height  floors_total  living_area  floor  is_apartment  studio  \
0         2.650391          16.0    51.000000      8         False   False   
1         2.650391          11.0    18.593750      1         False   False   
2         2.650391           5.0    34.312500      4         False   False   
3         2.650391 

Также из информации о колонках можно сделать следующие выводы:
1. в колонке total_images тип int64, хотя диапазон чисел в этой колонке от 0 до 50 - можем изменить тип на int8
2. в колонке total_area тип float64, хотя диапазон чисел в этой колонке от 12 до 900 - можем изменить тип на float16
3. в колонке first_day_exposition необходимо преобразовать данные в datatype
4. в колонке rooms тип int64, хотя диапазон чисел в этой колонке от 0 до 19 - можем изменить тип на int8, также подозрительным выглядит значение 0 в данной колонке
5. в колонке ceiling_height тип float64, хотя диапазон чисел в этой колонке от 1 до 100 - можем изменить тип на float8, также подозрительным выглядит значение 1 в данной колонке
6. в колонке floors_total тип float64, хотя диапазон чисел в этой колонке от 1 до 60 и хранятся по факту целочисленные значения - можем изменить тип на int8
7. в колонке living_area тип float64, хотя диапазон чисел в этой колонке от 2 до 409.7 - можем изменить тип на float16
8. в колонке floor тип int64, хотя диапазон чисел в этой колонке от 1 до 33 - можем изменить тип на int8
9. в колонке is_apartment тип object, хотя в колонке хранятся булевы значения - можем изменить тип на bool
10. в колонке kitchen_area тип float64, хотя диапазон чисел в этой колонке от 1.3 до 112 - можем изменить тип на float8
11. в колонке balcony тип float64, хотя диапазон чисел в этой колонке от 0 до 5 и хранятся по факту целочисленные значения - можем изменить тип на int8
12. в колонке locality_name данные необходимо привести к нижнему регистру
13. в колонке airports_nearest тип float64, хотя диапазон чисел в этой колонке от 0 до 84869 - можем изменить тип на float32, также подозрительным выглядит значение 0 в данной колонке
14. в колонке cityCenters_nearest тип float64, хотя диапазон чисел в этой колонке от 181 до 65968 - можем изменить тип на float32, а также изменим название колонки на city_centers_nearest
15. в колонке parks_around3000 тип float64, хотя диапазон чисел в этой колонке от 0 до 3 и хранятся по факту целочисленные значения - можем изменить тип на int8, а также изменим название колонки на parks_around_3000
16. в колонке parks_nearest тип float64, хотя диапазон чисел в этой колонке от 1 до 31900 и хранятся по факту целочисленные значения - можем изменить тип на int8
17. в колонке ponds_around3000 тип float64, хотя диапазон чисел в этой колонке от 0 до 3 и хранятся по факту целочисленные значения - можем изменить тип на int8, а также изменим название колонки на ponds_around_3000
18. в колонке ponds_nearest тип float64, хотя диапазон чисел в этой колонке от 13 до 1344 - можем изменить тип на float16
19. в колонке days_exposition тип float64, хотя диапазон чисел в этой колонке от 1 до 15800 и хранятся по факту целочисленные значения - можем изменить тип на int16


In [117]:
# преобразуем типы колонок
data = data.astype(
    {
        'total_images': 'int8',
        'total_area': 'float16',
        'rooms': 'int8',
        'ceiling_height': 'float16',
        'floors_total': 'float16',
        'living_area': 'float16',
        'floor': 'int8',
        'is_apartment': 'bool',
        'kitchen_area': 'float16',
        'balcony': 'int8',
        'airports_nearest': 'float32',
        'cityCenters_nearest': 'float32',
        'parks_around3000': 'float16',
        'parks_nearest': 'float16',
        'ponds_around3000': 'float16',
        'ponds_nearest': 'float16',
        'days_exposition': 'float16'
    }
)
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23699 entries, 0 to 23698
Data columns (total 22 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   total_images          23699 non-null  int8   
 1   last_price            23699 non-null  float64
 2   total_area            23699 non-null  float16
 3   first_day_exposition  23699 non-null  object 
 4   rooms                 23699 non-null  int8   
 5   ceiling_height        23699 non-null  float16
 6   floors_total          23613 non-null  float16
 7   living_area           21796 non-null  float16
 8   floor                 23699 non-null  int8   
 9   is_apartment          23699 non-null  bool   
 10  studio                23699 non-null  bool   
 11  open_plan             23699 non-null  bool   
 12  kitchen_area          21421 non-null  float16
 13  balcony               23699 non-null  int8   
 14  locality_name         23650 non-null  object 
 15  airports_nearest   

In [17]:
len(data.query("parks_around3000.isna() or parks_around3000 == 0"))

15624

In [9]:
data.query("floors_total.isna()").count()

total_images            86
last_price              86
total_area              86
first_day_exposition    86
rooms                   86
ceiling_height           9
floors_total             0
living_area             53
floor                   86
is_apartment             0
studio                  86
open_plan               86
kitchen_area            39
balcony                 47
locality_name           85
airports_nearest        76
cityCenters_nearest     76
parks_around3000        76
parks_nearest           34
ponds_around3000        76
ponds_nearest           49
days_exposition         77
dtype: int64

In [18]:
len(data.query("parks_nearest.isna()"))

15620

In [32]:
print(f'int8: {2**8}; int16: {2**16}, int32: {2**32}, int64: {2**64}')

int8: 256; int16: 65536, int32: 4294967296, int64: 18446744073709551616
